internal/server/singleprocess/service_trigger.go

```diff
@@ -71,3 +75,166 @@ func (s *service) ListTriggers(
+			}
+		}
+
+		job.Variables = req.VariableOverrides
+	}
+
+	// TODO(briancain): look up a target runner config at the project/app level and apply it to job requests
+	job.TargetRunner = &pb.Ref_Runner{Target: &pb.Ref_Runner_Any{}}
+
+	// generate job requests
+	var jobList []*pb.QueueJobRequest
+	var ids []string
+	if runTrigger.Application == nil || runTrigger.Application.Application == "" {
+		// we're gonna queue multiple jobs for every application in a project
+		log.Debug("building multi-jobs for all apps in project", "project", runTrigger.Project.Project)
+		jobList, err = s.state.JobProjectScopedRequest(runTrigger.Project, job)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		log.Debug("building a single job for target", "project",
+			runTrigger.Application.Project, "app", runTrigger.Application.Application)
+		// we're only targetting a specific application, so queue 1 job
+		job.Application = runTrigger.Application
+		j := &pb.QueueJobRequest{Job: job}
+		jobList = append(jobList, j)
+	}
+
+	// NOTE(briancain): See https://github.com/hashicorp/waypoint/issues/2884
+	// for why we must attach the full PushedArtifact message for a deploy operation
+	if deployArtifactSet {
+		// We have to set the full artifact message on Deployment operations
+		for i, qJob := range jobList {
+			switch op := qJob.Job.Operation.(type) {
+			case *pb.Job_Deploy:
+				if op.Deploy.Artifact == nil {
+					// get latest pushed artifact, then set it on the operation
+					artifactLatest, err := s.state.ArtifactLatest(qJob.Job.Application, qJob.Job.Workspace)
+					if err != nil {
+						return nil, status.Errorf(codes.Internal, "failed to obtain latest pushed artifact: %s", err)
+					}
+
+					jobList[i].Job.Operation = &pb.Job_Deploy{
+						Deploy: &pb.Job_DeployOp{
+							Artifact: artifactLatest,
+						},
+					}
+				} else {
+					// Set the actual pushed artifact on the operation
+					buildSeq := op.Deploy.Artifact.Sequence
+					artifact, err := s.state.ArtifactGet(&pb.Ref_Operation{
+						Target: &pb.Ref_Operation_Sequence{
+							Sequence: &pb.Ref_OperationSeq{
+								Application: qJob.Job.Application,
+								Number:      buildSeq,
+							},
+						},
+					})
+					if err != nil {
+						return nil, status.Errorf(codes.Internal, "failed to obtain pushed artifact id %q: %s", buildSeq, err)
+					}
+
+					jobList[i].Job.Operation = &pb.Job_Deploy{
+						Deploy: &pb.Job_DeployOp{
+							Artifact: artifact,
+						},
+					}
+				}
+			}
+		}
+	}
+
+	if len(jobList) > 0 {
+		// Queue the job(s)
+		log.Debug("queueing jobs", "total_jobs", len(jobList))
+
+		// NOTE(briancain): queueJobMulti currently returns 3 Jobs due to On-Demand Runners:
+		// The start task job, the actual queued job, and the stop task job. Users
+		// will generally just care about the middle job for streaming the result
+		// of the RunTrigger
+		respList, err := s.queueJobMulti(ctx, jobList)
```

Comment:
It's probably ok to leave this out for now, but we should figure out if we should wait for the jobs to be accepted (or at least one job because of ODR) before returning, so that the caller knows that at least the job was valid when the RPC call returns.

Reply:
That's a good idea. I ran into some issues while testing where I'd run a bunch of triggers, get ids back, and not see the actual job errors until I looked at the server logs much later. I'll see what I can do to make it better this week before release.

Reply:
Discussed via slack, these jobs are running through on-demand runners, which validate the job already. So no need to do it again here!